kafka-connect/src/main/java/io/tabular/iceberg/connect/channel/CommitterImpl.java

```diff
@@ -0,0 +1,252 @@
+  private final UUID producerId;
+  private final Producer<String, byte[]> producer;
+
+  public CommitterImpl(SinkTaskContext context, IcebergSinkConfig config) {
+    this(
+        context,
+        config,
+        new ConsumerFactoryImpl(),
+        new AdminFactoryImpl(),
+        new ProducerFactoryImpl());
+  }
+
+  private CommitterImpl(
+      SinkTaskContext context,
+      IcebergSinkConfig config,
+      ConsumerFactory consumerFactory,
+      AdminFactory adminFactory,
+      ProducerFactory producerFactory) {
+    this(
+        context,
+        config,
+        new ControlTopicCommitRequestListener(config, consumerFactory),
+        adminFactory,
+        producerFactory,
+        new CoordinatorThreadFactoryImpl(adminFactory, consumerFactory, producerFactory));
+  }
+
+  @VisibleForTesting
+  CommitterImpl(
+      SinkTaskContext context,
+      IcebergSinkConfig config,
+      CommitRequestListener commitRequestListener,
+      AdminFactory adminFactory,
+      ProducerFactory producerFactory,
+      CoordinatorThreadFactory coordinatorThreadFactory) {
+    this.context = context;
+    this.config = config;
+    this.isLeader = Objects.equals(config.taskId(), 0);
+
+    if (isLeader) {
+      this.coordinatorThread = coordinatorThreadFactory.create(config);
+    } else {
+      this.coordinatorThread = null;
+    }
+
+    this.commitRequestListener = commitRequestListener;
+
+    Map<String, String> producerProps = Maps.newHashMap(config.kafkaProps());
+    // use a random transactional-id to avoid producer generation based fencing
+    producerProps.put(
+        ProducerConfig.TRANSACTIONAL_ID_CONFIG,
+        String.join(
+            "-",
+            config.connectGroupId(),
+            config.taskId().toString(),
+            "committer",
+            UUID.randomUUID().toString()));
+    Pair<UUID, Producer<String, byte[]>> pair = producerFactory.create(producerProps);
+    this.producerId = pair.first();
+    this.producer = pair.second();
+    producer.initTransactions();
+
+    // The source-of-truth for source-topic offsets is the control-group-id
+    Map<TopicPartition, Long> committedOffsets =
+        committedOffsets(adminFactory, config.controlGroupId());
+    // Rewind kafka connect consumer to avoid duplicates
+    context.offset(committedOffsets);
+  }
+
+  private Map<TopicPartition, Long> committedOffsets(AdminFactory adminFactory, String groupId) {
+    try (Admin admin = adminFactory.create(config.kafkaProps())) {
+      ListConsumerGroupOffsetsResult response =
+          admin.listConsumerGroupOffsets(
+              groupId, new ListConsumerGroupOffsetsOptions().requireStable(true));
+      return response.partitionsToOffsetAndMetadata().get().entrySet().stream()
+          .filter(entry -> context.assignment().contains(entry.getKey()))
+          .collect(toMap(Map.Entry::getKey, entry -> entry.getValue().offset()));
+    } catch (InterruptedException | ExecutionException e) {
```

Comment:
nit/question: Is this wrapping/rethrowing necessary?  Does Kafka Connect do something special w/ ConnectException via control flow?

Reply:
> Does Kafka Connect do something special w/ ConnectException via control flow?

Not that I know of. I've just copy-pasted the original code into here, trying to minimize unnecessary changes for this refactor.